src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs

```diff
@@ -923,36 +926,52 @@ private Type FindType(int typeIndex)
             }
             if (_typeTable[typeIndex] == null)
             {
-                long oldPos = _store.BaseStream.Position;
-                try
-                {
-                    _store.BaseStream.Position = _typeNamePositions[typeIndex];
-                    string typeName = _store.ReadString();
-                    _typeTable[typeIndex] = Type.GetType(typeName, true);
-                }
-                // If serialization isn't supported, we convert FileNotFoundException to
-                // NotSupportedException for consistency with v2. This is a corner-case, but the
-                // idea is that we want to give the user a more accurate error message. Even if
-                // the dependency were found, we know it will require serialization since it
-                // can't be one of the types we special case. So if the dependency were found,
-                // it would go down the serialization code path, resulting in NotSupported for
-                // SKUs without serialization.
-                //
-                // We don't want to regress the expected case by checking the type info before
-                // getting to Type.GetType -- this is costly with v1 resource formats.
-                catch (FileNotFoundException)
+                if (AllowCustomResourceTypes)
                 {
-                    throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+                    UseReflectionToGetType(typeIndex);
                 }
-                finally
+                else
                 {
-                    _store.BaseStream.Position = oldPos;
+                    throw new NotSupportedException(SR.ResourceManager_ReflectionNotAllowed);
```

Comment:
Can we throw as the very first thing in this method? It would allow a bit more code to be trimmed. E.g. it can be something like:

```
private Type FindType(int typeIndex)
{
    if (!AllowCustomResourceTypes)
        throw new NotSupportedException(SR.ResourceManager_ReflectionNotAllowed);

    if (typeIndex < 0 || typeIndex >= _typeTable.Length)
        throw new BadImageFormatException(SR.BadImageFormat_InvalidType);

    return _typeTable[typeIndex] ?? UseReflectionToGetType(typeIndex);
}
```